Sources/_SubprocessCShims/process_shims.c

```diff
@@ -303,154 +265,100 @@ int _subprocess_spawn(
+    char * _Nonnull str2
 ) {
-    // Setup stdin, stdout, and stderr
-    posix_spawn_file_actions_t file_actions;
-
-    int rc = posix_spawn_file_actions_init(&file_actions);
-    if (rc != 0) { return rc; }
-    if (file_descriptors[0] >= 0) {
-        rc = posix_spawn_file_actions_adddup2(
-            &file_actions, file_descriptors[0], STDIN_FILENO
-        );
-        if (rc != 0) { return rc; }
-    }
-    if (file_descriptors[2] >= 0) {
-        rc = posix_spawn_file_actions_adddup2(
-            &file_actions, file_descriptors[2], STDOUT_FILENO
-        );
-        if (rc != 0) { return rc; }
-    }
-    if (file_descriptors[4] >= 0) {
-        rc = posix_spawn_file_actions_adddup2(
-            &file_actions, file_descriptors[4], STDERR_FILENO
-        );
-        if (rc != 0) { return rc; }
-    }
-    // Setup working directory
-    if (working_directory != NULL) {
-        rc = _subprocess_addchdir_np(&file_actions, working_directory);
-        if (rc != 0) {
-            return rc;
+    return snprintf(str, len, format, str1, str2);
+}
+
+static int _positive_int_parse(const char *str) {
+    int out = 0;
+    char c = 0;
+
+    while ((c = *str++) != 0) {
+        out *= 10;
+        if (c >= '0' && c <= '9') {
+            out += c - '0';
+        } else {
+            return -1;
         }
     }
+    return out;
+}
 
-    // Close parent side
-    if (file_descriptors[1] >= 0) {
-        rc = posix_spawn_file_actions_addclose(&file_actions, file_descriptors[1]);
-        if (rc != 0) { return rc; }
-    }
-    if (file_descriptors[3] >= 0) {
-        rc = posix_spawn_file_actions_addclose(&file_actions, file_descriptors[3]);
-        if (rc != 0) { return rc; }
-    }
-    if (file_descriptors[5] >= 0) {
-        rc = posix_spawn_file_actions_addclose(&file_actions, file_descriptors[5]);
-        if (rc != 0) { return rc; }
+static int _highest_possibly_open_fd_dir(const char *fd_dir) {
+    int highest_fd_so_far = 0;
+    DIR *dir_ptr = opendir(fd_dir);
+    if (dir_ptr == NULL) {
+        return -1;
     }
 
-    // Setup spawnattr
-    posix_spawnattr_t spawn_attr;
-    rc = posix_spawnattr_init(&spawn_attr);
-    if (rc != 0) { return rc; }
-    // Masks
-    sigset_t no_signals;
-    sigset_t all_signals;
-    sigemptyset(&no_signals);
-    sigfillset(&all_signals);
-    rc = posix_spawnattr_setsigmask(&spawn_attr, &no_signals);
-    if (rc != 0) { return rc; }
-    rc = posix_spawnattr_setsigdefault(&spawn_attr, &all_signals);
-    if (rc != 0) { return rc; }
-    // Flags
-    short flags = POSIX_SPAWN_SETSIGMASK | POSIX_SPAWN_SETSIGDEF;
-    if (process_group_id != NULL) {
-        flags |= POSIX_SPAWN_SETPGROUP;
-        rc = posix_spawnattr_setpgroup(&spawn_attr, *process_group_id);
-        if (rc != 0) { return rc; }
+    struct dirent *dir_entry = NULL;
+    while ((dir_entry = readdir(dir_ptr)) != NULL) {
+        char *entry_name = dir_entry->d_name;
+        int number = _positive_int_parse(entry_name);
```

Comment:
Is this really the right way to do this? Seems very ad-hoc.

Reply:
(and racy)

Reply:
Yes. So there are generally two ways to determine the highest fd number on Linux:
1) Use `getrlimit`
2) Read `/proc/self/fd`
The difference between the two is that 1) only tells you the limit whereas 2) tells you the actual highest fd opened. I decided to go for 2) because it's more accurate. Also since this code is executed between `fork()` and `exec()`, there is no threading issues here.

Reply:
@iCharlesHu I implemented a version of this in `AsyncProcess` which should be (v)fork safe because it doesn't allocate: https://github.com/swiftlang/swift-sdk-generator/blob/3f4a6a800414010da0057985f80b2aba5dd4fc60/Sources/CProcessSpawnSync/internal-helpers.h#L57-L98

(glibc (LGPL!) also has an implementation: https://github.com/bminor/glibc/blob/6afece738c2b8408585272a95090ce5d5345dd19/sysdeps/unix/sysv/linux/closefrom_fallback.c#L31)